x/exchange/keeper/payments.go

```diff
@@ -0,0 +1,482 @@
+	if err != nil {
+		return fmt.Errorf("error marshaling payment: %w", err)
+	}
+
+	var target sdk.AccAddress
+	var iKey []byte
+	if len(payment.Target) > 0 {
+		target, err = sdk.AccAddressFromBech32(payment.Target)
+		if err != nil {
+			return fmt.Errorf("invalid target %q: %w", payment.Target, err)
+		}
+		iKey = MakeIndexKeyTargetToPayment(target, source, payment.ExternalId)
+	}
+
+	var oldIKey []byte
+	if existing, _ := k.getPaymentFromStore(store, source, payment.ExternalId); existing != nil {
+		switch existing.Target {
+		case "":
+			// There isn't an entry yet, so there's nothing to delete.
+		case payment.Target:
+			// The existing entry has the same target. No need to delete it and rewrite the same index entry.
+			iKey = nil
+		default:
+			// The target's changing, delete the old index entry.
+			var oldTarget sdk.AccAddress
+			oldTarget, err = sdk.AccAddressFromBech32(existing.Target)
+			if err == nil && len(oldTarget) > 0 {
+				oldIKey = MakeIndexKeyTargetToPayment(oldTarget, source, payment.ExternalId)
+			}
+		}
+	}
+
+	store.Set(pKey, pVal)
+	if len(oldIKey) > 0 {
+		store.Delete(oldIKey)
+	}
+	if len(iKey) > 0 {
+		store.Set(iKey, []byte{})
+	}
+
+	return nil
+}
+
+// createPaymentInStore verifies that the provided payment does not yet exist, then writes it to the state store.
+func (k Keeper) createPaymentInStore(store sdk.KVStore, payment *exchange.Payment) error {
+	if paymentExists(store, payment.Source, payment.ExternalId) {
+		return fmt.Errorf("a payment already exists with source %s and external id %q",
+			payment.Source, payment.ExternalId)
+	}
+	return k.setPaymentInStore(store, payment)
+}
+
+// deletePaymentFromStore deletes a payment (and its index) from the state store.
+func deletePaymentFromStore(store sdk.KVStore, payment *exchange.Payment) error {
+	if payment == nil {
+		return errors.New("cannot delete nil payment")
+	}
+
+	source, err := sdk.AccAddressFromBech32(payment.Source)
+	if err != nil {
+		return fmt.Errorf("invalid source %q: %w", payment.Source, err)
+	}
+	pKey := MakeKeyPayment(source, payment.ExternalId)
+
+	var target sdk.AccAddress
+	var iKey []byte
+	if len(payment.Target) > 0 {
+		target, err = sdk.AccAddressFromBech32(payment.Target)
+		if err != nil {
+			return fmt.Errorf("invalid target %q: %w", payment.Target, err)
+		}
+		iKey = MakeIndexKeyTargetToPayment(target, source, payment.ExternalId)
+	}
+
+	store.Delete(pKey)
+	if len(iKey) > 0 {
+		store.Delete(iKey)
+	}
+
+	return nil
+}
+
+// deletePaymentAndReleaseHold deletes a payment from the state store and releases its hold.
+func (k Keeper) deletePaymentAndReleaseHold(ctx sdk.Context, store sdk.KVStore, payment *exchange.Payment) error {
+	err := deletePaymentFromStore(store, payment)
+	if err != nil {
+		if payment != nil {
+			return fmt.Errorf("error deleting payment with source %s and external id %q: %w", payment.Source, payment.ExternalId, err)
+		}
+		return fmt.Errorf("error deleting payment: %w", err)
+	}
```

Comment:
I'm not sure if you care about consistency, but you can make this more consistent with your other functions by moving line 174 into its own nil check. I understand deletePaymentFromStore does this check so its up to you.

```
if payment == nil {
  return fmt.Errorf("error deleting payment and release hold on nil payment")
}
err := deletePaymentFromStore(store, payment)
if err != nil {
  return fmt.Errorf("error deleting payment with source %s and external id %q: %w", payment.Source, payment.ExternalId, err)
}
```

Reply:
Resolving since it's not necessary.